cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -424,14 +444,99 @@ func (c ControllerContext) IsControllerEnabled(controllerDescriptor *ControllerD
 	return genericcontrollermanager.IsControllerEnabled(controllerDescriptor.Name(), controllersDisabledByDefault, c.ComponentConfig.Generic.Controllers)
 }
 
-// InitFunc is used to launch a particular controller. It returns a controller
-// that can optionally implement other interfaces so that the controller manager
-// can support the requested features.
-// The returned controller may be nil, which will be considered an anonymous controller
-// that requests no additional features from the controller manager.
-// Any error returned will cause the controller process to `Fatal`
-// The bool indicates whether the controller was enabled.
-type InitFunc func(ctx context.Context, controllerContext ControllerContext, controllerName string) (controller controller.Interface, enabled bool, err error)
+// NewClientConfig is a shortcut for ClientBuilder.Config. It wraps the error with an additional message.
+func (c ControllerContext) NewClientConfig(name string) (*restclient.Config, error) {
+	config, err := c.ClientBuilder.Config(name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kubernetes client config for %q: %w", name, err)
+	}
+	return config, nil
+}
+
+// NewClient is a shortcut for ClientBuilder.Client. It wraps the error with an additional message.
+func (c ControllerContext) NewClient(name string) (kubernetes.Interface, error) {
+	client, err := c.ClientBuilder.Client(name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kubernetes client for %q: %w", name, err)
+	}
+	return client, nil
+}
+
+// Runnable represents a loop implementing any piece of logic really.
+type Runnable interface {
+	// Run implements any kind of processing loop.
+	// When there is anything to be done, it blocks until the context is cancelled.
+	// Run must ensure all goroutines are terminated before returning.
+	Run(context.Context)
+}
+
+// Controller defines the base interface that all controller wrappers must implement.
+type Controller interface {
+	// Name returns the controller's canonical name.
+	Name() string
+
+	// Runnable is the main controller loop.
+	Runnable
+}
+
+// runnableFunc makes it possible to turn a function into a Runnable quickly.
+type runnableFunc func(context.Context)
+
+func (run runnableFunc) Run(ctx context.Context) {
+	run(ctx)
+}
+
+// runnables is a slice of Runnable objects that are to be run concurrently.
+type runnables []Runnable
+
+func (rx runnables) Run(ctx context.Context) {
+	var wg sync.WaitGroup
+	wg.Add(len(rx))
+	for _, r := range rx {
+		go func() {
+			defer wg.Done()
+			r.Run(ctx)
+		}()
+	}
+	wg.Wait()
+}
+
+func runRunnables(ctx context.Context, rx ...Runnable) {
+	runnables(rx).Run(ctx)
+}
+
+// named can be easily embedded to implement Name() for a struct.
+type named string
```

Comment:
This section feels a bit overkill imo, and we have to switch from string to named. Can we just use a named struct for the named runnable rather than declaring another type? I get the point is so that we don't have to redeclare the run function but I think that's more clear than an anonymous struct.

```
func (n namedRunnable) Name() string {
	return n.name
}

func (n namedRunnable) Run(ctx context.Context) {
	n.runnable.Run(ctx)
}

// namedRunnable implements the Controller interface.
type namedRunnable struct {
	name     string
	runnable Runnable
}
```

Reply:
Aligned in 2499352caae560fb0811ead4ef089408b51372d8 . I kept `Runnable` embedded, though. It was like that before I decided to overengineer this 😅 Let me know if you strongly feel that there should be no embedding.

Reply:
👍 makes sense to me.